### PR TITLE
support OAuth for pulsar scaler

### PR DESCRIPTION
**Pull request by mingmcb**
support OAuth for pulsar scaler
### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Relates to #4700

Relates to https://github.com/kedacore/keda-docs/pull/1161

### CODE REVIEW

**Comment by mingmcb** [comment]
The PR for document is https://github.com/kedacore/keda-docs/pull/1161

**Review by zroubalik — COMMENTED** [review]
Looking great, my only concern is about exposing auth data in the trigger metadata, see: https://github.com/kedacore/keda-docs/pull/1161#pullrequestreview-1490251516

File: pkg/scalers/authentication/authentication_helpers.go [hunk 1]
```diff
@@ -80,6 +85,20 @@ func GetAuthConfigs(triggerMetadata, authParams map[string]string) (out *AuthMet
...
+		case OAuthType:
+			if out.EnableBasicAuth {
+				return nil, errors.New("both oauth and basic authentication can not be set")
+			}
+			if out.EnableBearerAuth {
+				return nil, errors.New("both oauth and bearer authentication can not be set")
+			}
+			out.EnableOAuth = true
+			out.OauthTokenURI = authParams["oauthTokenURI"]
+			if authParams["scope"] != "" {
```
Comment (zroubalik) [inline pkg/scalers/authentication/authentication_helpers.go:97]: could you please trim the string first and change the format to accept **a comma separated list* of scopes?  To be consistent with the rest of the config.

File: pkg/scalers/pulsar_scaler.go [hunk 2]
```diff
@@ -199,6 +200,18 @@ func parsePulsarMetadata(config *ScalerConfig) (pulsarMetadata, error) {
 	if err != nil {
 		return meta, fmt.Errorf("error parsing %s: %w", msgBacklogMetricName, err)
 	}
+
+	if auth != nil && auth.EnableOAuth {
+		if auth.OauthTokenURI == "" {
+			auth.OauthTokenURI = config.TriggerMetadata["oauthTokenURI"]
+		}
+		if auth.Scopes == nil && config.TriggerMetadata["scope"] != "" {
+			auth.Scopes = strings.Split(config.TriggerMetadata["scope"], " ")
```
Comment (zroubalik) [inline pkg/scalers/pulsar_scaler.go:209]: could you please trim the string first and change the format to accept *a comma separated list of scopes? To be consistent with the rest of the config.

**Review by zroubalik — CHANGES_REQUESTED** [review]
please change the format for `scope`

File: CHANGELOG.md [hunk 3]
```diff
@@ -105,6 +105,7 @@ New deprecation(s):
 - **RabbitMQ Scaler**: Add support for `unsafeSsl` in trigger metadata ([#4448](https://github.com/kedacore/keda/issues/4448))
 - **RabbitMQ Scaler**: Add support for `workloadIdentityResource` and utilize AzureAD Workload Identity for HTTP authorization ([#4716](https://github.com/kedacore/keda/issues/4716))
 - **Solace Scaler**: Add new `messageReceiveRateTarget` metric to Solace Scaler ([#4665](https://github.com/kedacore/keda/issues/4665))
+- **Pulsar Scaler**: Add support for OAuth extensions ([#4700](https://github.com/kedacore/keda/issues/4700))
```
Comment (zroubalik) [inline CHANGELOG.md:108]: Please move this to unreleased section

File: pkg/scalers/authentication/authentication_helpers.go [hunk 4]
```diff
@@ -80,6 +85,21 @@ func GetAuthConfigs(triggerMetadata, authParams map[string]string) (out *AuthMet
...
+		case OAuthType:
+			if out.EnableBasicAuth {
+				return nil, errors.New("both oauth and basic authentication can not be set")
+			}
+			if out.EnableBearerAuth {
+				return nil, errors.New("both oauth and bearer authentication can not be set")
+			}
+			out.EnableOAuth = true
+			out.OauthTokenURI = authParams["oauthTokenURI"]
+			scope := strings.ReplaceAll(authParams["scope"], " ", "")
```
Comment (zroubalik) [inline pkg/scalers/authentication/authentication_helpers.go:97]: isn't `strings.TrimSpace()` better?

File: pkg/scalers/pulsar_scaler.go [hunk 5]
```diff
@@ -199,6 +200,19 @@ func parsePulsarMetadata(config *ScalerConfig) (pulsarMetadata, error) {
 	if err != nil {
 		return meta, fmt.Errorf("error parsing %s: %w", msgBacklogMetricName, err)
 	}
+
+	if auth != nil && auth.EnableOAuth {
+		if auth.OauthTokenURI == "" {
+			auth.OauthTokenURI = config.TriggerMetadata["oauthTokenURI"]
+		}
+		scope := strings.ReplaceAll(config.TriggerMetadata["scope"], " ", "")
```
Comment (zroubalik) [inline pkg/scalers/pulsar_scaler.go:208]: isn't `strings.TrimSpace()` better?

**Comment by zroubalik** [comment]
This is great, there's also a proposal for a generic OAuth support that I like. I wonder if we can go this direction as well for Pulsar? https://github.com/kedacore/keda/issues/4694

File: CHANGELOG.md [hunk 6]
```diff
@@ -45,6 +45,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - [v1.0.0](#v100)
 
 ## Unreleased
+- **Pulsar Scaler**: Add support for OAuth extensions ([#4700](https://github.com/kedacore/keda/issues/4700))
```
Comment (zroubalik) [inline CHANGELOG.md:48]: Please move it to Improvements subsection.

File: pkg/scalers/authentication/authentication_helpers.go [hunk 7]
```diff
@@ -92,6 +109,25 @@ func GetAuthConfigs(triggerMetadata, authParams map[string]string) (out *AuthMet
 	return out, err
 }
 
+func ParseScope(inputStr string) []string {
```
Comment (zroubalik) [inline pkg/scalers/authentication/authentication_helpers.go:112]: Could you please add a short describtion of what this function does?

```go
// ParseScope ...
```

**Review by zroubalik — COMMENTED** [review]
Looking good, minor stuff.

**Comment by zroubalik** [comment]
> Looking great, my only concern is about exposing auth data in the trigger metadata, see: [kedacore/keda-docs#1161 (review)](https://github.com/kedacore/keda-docs/pull/1161#pullrequestreview-1490251516)

@kedacore/keda-core-contributors WDYT?

**Comment by mingmcb** [comment]
> This is great, there's also a proposal for a generic OAuth support that I like. I wonder if we can go this direction as well for Pulsar? #4694

Doesn't look like #4694 is implemented, I can convert to use it if it is ready.

**Review by zroubalik — COMMENTED** [review]
@mingmcb do you think you can address the review comments? We can proceed and merge the PR then.

We are back after a summer break

File: pkg/scalers/pulsar_scaler.go [hunk 8]
```diff
@@ -5,15 +5,16 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
-	"io"
-	"net/http"
-	"strconv"
-	"strings"
-
 	"github.com/go-logr/logr"
+	"golang.org/x/oauth2/clientcredentials"
```
Comment (zroubalik) [inline pkg/scalers/pulsar_scaler.go:9]: please keep imports separated in these 3 groups: 
- internal imports 
- external libraries
- KEDA packages

**Review by zroubalik — APPROVED** [review]
LGTM

**Review by JorTurFer — APPROVED** [review]
LGTM

**Review by JorTurFer — APPROVED** [review]
LGTM

**Pull request merged**